### PR TITLE
IRSA-1016: exposes workspace components to JS API

### PR DESCRIPTION
**Pull request by loitly**
https://jira.ipac.caltech.edu/browse/IRSA-1016

Use table test page to test this feature.  Make sure to use your fully qualify host name in the url.
Log into irsa.ipac.caltech.edu first.
http://facade.ipac.caltech.edu:8080/firefly/demo/ffapi-table-test.html

You'll need to build locally with `ignore.auth=false`.  
You can do this by adding `-Pignore.auth=false` to the gradle command line.
Or, by inserting that line into your `~/.gradle/build.config` file to use that option on every build.

### CODE REVIEW

**Comment by ejoliet** [comment]
I've logged in irsa but when i click on the workspace button it says that i don't have access and need to login.

**Comment by loitly** [comment]
> I've logged in irsa but when i click on the workspace button it says that i don't have access and need to login.

Did you build with `-Pignore.auth=false`?  Also, you should log into `irsa ops` and use FQN for your own host.

**Comment by ejoliet** [comment]
I forgot about FQN but i can't get to any tomcat pages with my full qualified name...

**Comment by loitly** [comment]
Okay, you're on wifi.  You need to be on LAN.

**Comment by loitly** [comment]
> Error:
> Column not found: clon"

You are seeing this error because that test page has a sort by `clon` parameter in it.  The document you selected must not have that column.  The error message in my opinion is valid.

File: src/firefly/js/ui/WorkspaceViewer.jsx [hunk 1]
```diff
@@ -144,93 +112,132 @@ WorkspaceView.propTypes = {
...
- * @returns {XML}
- * @constructor
+ * Show WorkspaceView as a pop up field.  A button is used to trigger the popup.
+ * A label next to the button to show the selected item.
  */
-function WorkspaceReadView({wrapperStyle, onClickUpload, value, isLoading}) {
+export const WorkspacePickerPopup =  fieldGroupConnector(WorkspaceAsPopup,
+                                        ({fieldKey, onComplete, value={}, ...rest}, fireValueChange) => {
+                                            const onSelComplete = (v) => {
+                                                fireValueChange({value: v});
```
Comment (cwang2016) [inline src/firefly/js/ui/WorkspaceViewer.jsx]: should this be v[fieldKey]?

File: src/firefly/js/ui/WorkspaceViewer.jsx [hunk 2]
```diff
@@ -247,7 +254,7 @@ function getUploadProps(params, fireValueChange) {
     return Object.assign({}, params,
         {
             value: params.displayValue,
-            onClickUpload: resultSuccess(fireValueChange, params.fileAnalysis, params.preloadWsFile)
+            onComplete: resultSuccess(fireValueChange, params.fileAnalysis, params.preloadWsFile)
```
Comment (cwang2016) [inline src/firefly/js/ui/WorkspaceViewer.jsx]: add 'fieldKey', 
resultSuccess(fireValueChange, params.fileAnalysis, params.preloadWsFile, params.fieldKey)

File: src/firefly/js/ui/WorkspaceViewer.jsx [hunk 3]
```diff
@@ -481,4 +401,113 @@ export function workspacePopupMsg(msg, title) {
...
+                    </FieldGroup>
+
+                    <div style={{display: 'flex', justifyContent: 'space-between',
+                                 margin: `20px ${HMargin}px ${VMargin}px ${HMargin}px`}}>
+                        <div style={{display: 'flex', width: '60%', alignItems: 'flex-end'}}>
+                            <div style={{marginRight: 10}}>
+                                <CompleteButton
+                                    groupKey={workspacePopupGroup}
+                                    onSuccess={onComplete}
+                                    onFail={resultFail()}
```
Comment (cwang2016) [inline src/firefly/js/ui/WorkspaceViewer.jsx]: resultFail(fieldKey)

**Comment by cwang2016** [comment]
workspace upload is broken using <WorkspaceUpload> component due to wrong value of 'fieldKey'. Here is suggestion for the fixing,  
- passing fieldKey to function resultSuccess() and redultFail() in 'WorkspaceViewer.jsx'  like 
`resultSuccess(fireValueChange, fileAnalysis, preloadWsFile, fieldKey=workspaceUploadDef.file.fkey)
      resultFail(fieldKey=workspaceUploadDef.file.fkey)`
  and changing 'workspaceUploadDef.file.fkey' in these functions to be 'fieldKey'

One concern regarding workspace upload component made for API: when the workspace upload button is clicked, is the function that preloading the file from the workspace considered? or only getting the workspace file name from the filepicker is considered?

**Comment by ejoliet** [comment]
>  is the function that preloading the file from the workspace considered? or only getting the workspace file name from the filepicker is considered?

 me it makes more sense to pass back the url of the file to the client and avoid preloading it but does this make sens in Gator context or any other use case?

**Pull request merged**